crates/bevy_ecs/hecs/src/dynamic_query.rs

```diff
@@ -0,0 +1,527 @@
+                matches_any = true;
+
+                fetch
+                    .immutable
+                    .push(NonNull::new_unchecked(ptr.as_ptr().add(offset)));
+            } else {
+                missing_any = true;
+            }
+        }
+
+        // Query mutable components
+        for component in &self.mutable {
+            let ptr = archetype.get_dynamic(component.id, component.layout.size(), 0);
+
+            if let Some(ptr) = ptr {
+                matches_any = true;
+
+                fetch
+                    .mutable
+                    .push(NonNull::new_unchecked(ptr.as_ptr().add(offset)));
+            } else {
+                missing_any = true;
+            }
+        }
+
+        if matches_any && !missing_any {
+            Some(fetch)
+        } else {
+            None
+        }
+    }
+}
+
+pub struct DynamicQueryFetch<'query> {
+    query: &'query DynamicQuery,
+    entity: Option<NonNull<Entity>>,
+    immutable: Vec<NonNull<u8>>,
+    mutable: Vec<NonNull<u8>>,
+}
+
+impl<'query> DynamicQueryFetch<'query> {
+    fn new(query: &'query DynamicQuery) -> Self {
+        Self {
+            query,
+            entity: Default::default(),
+            immutable: Default::default(),
+            mutable: Default::default(),
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Index must be within bounds
+    pub(crate) unsafe fn fetch<'a>(&mut self, index: usize) -> DynamicQueryResult<'a> {
+        let mut query_result = DynamicQueryResult {
+            // FIXME: Does this entity need to be cloned to be safe, or is it safe just to
+            // dereference it?
+            entity: self.entity.map(|x| *x.as_ptr().add(index)),
+            immutable: Vec::with_capacity(self.immutable.len()),
+            mutable: Vec::with_capacity(self.mutable.len()),
+        };
+
+        for (component, ptr) in self.query.immutable.iter().zip(&self.immutable) {
+            query_result.immutable.push(&*slice_from_raw_parts(
+                ptr.as_ptr().add(index * component.layout.size()),
+                component.layout.size(),
+            ));
+        }
+
+        for (component, ptr) in self.query.mutable.iter().zip(&self.mutable) {
+            query_result.mutable.push(&mut *slice_from_raw_parts_mut(
+                ptr.as_ptr().add(index * component.layout.size()),
+                component.layout.size(),
+            ));
+        }
+
+        query_result
+    }
+}
+
+/// The data returned by [`DynamicFetch::fetch`]
+#[derive(Default)]
+pub struct DynamicQueryResult<'a> {
+    /// The entity, if requested in the Fetch
+    pub entity: Option<Entity>,
+    /// The immutable data for the immutable components requested
+    pub immutable: Vec<&'a [u8]>,
+    /// The mutable data for the immutable components requested
+    pub mutable: Vec<&'a mut [u8]>,
+}
+
+/// Information about a dynamic component
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct DynamicComponentInfo {
```

Comment:
Is this intended to be used exclusively for dynamic components, i.e. ones not identified by a `TypeId`? If not, should `DynamicQuery` be extended to support those as well?

Reply:
Because `id` is a `ComponentId` it can represent queries to both rust components and dynamically added components with an external ID.